model/link_metadata.go

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package model
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"hash/fnv"
+	"net/http"
+	"time"
+
+	"github.com/dyatlov/go-opengraph/opengraph"
+)
+
+const (
+	LINK_METADATA_TYPE_IMAGE     LinkMetadataType = "image"
+	LINK_METADATA_TYPE_NONE      LinkMetadataType = "none"
+	LINK_METADATA_TYPE_OPENGRAPH LinkMetadataType = "opengraph"
+)
+
+type LinkMetadataType string
+
+// LinkMetadata stores arbitrary data about a link posted in a message. This includes dimensions of linked images
+// and OpenGraph metadata.
+type LinkMetadata struct {
+	// Hash is a value computed from the URL and Timestamp for use as a primary key in the database.
+	Hash int64
+
+	URL       string
+	Timestamp int64
+	Type      LinkMetadataType
+
+	// Data is the actual metadata for the link. It should contain data of one of the following types:
+	// - *model.PostImage if the linked content is an image
+	// - *opengraph.OpenGraph if the linked content is an HTML document
+	// - nil if the linked content has no metadata
+	Data interface{}
+}
+
+func (o *LinkMetadata) PreSave() {
+	o.Hash = GenerateLinkMetadataHash(o.URL, o.Timestamp)
+}
+
+func (o *LinkMetadata) IsValid() *AppError {
+	if o.URL == "" {
+		return NewAppError("LinkMetadata.IsValid", "model.link_metadata.is_valid.url.app_error", nil, "", http.StatusBadRequest)
+	}
+
+	if o.Timestamp == 0 || !isRoundedToNearestHour(o.Timestamp) {
+		return NewAppError("LinkMetadata.IsValid", "model.link_metadata.is_valid.timestamp.app_error", nil, "", http.StatusBadRequest)
+	}
+
+	switch o.Type {
+	case LINK_METADATA_TYPE_IMAGE:
+		if o.Data == nil {
+			return NewAppError("LinkMetadata.IsValid", "model.link_metadata.is_valid.data.app_error", nil, "", http.StatusBadRequest)
+		}
+
+		if _, ok := o.Data.(*PostImage); !ok {
+			return NewAppError("LinkMetadata.IsValid", "model.link_metadata.is_valid.data_type.app_error", nil, "", http.StatusBadRequest)
+		}
+	case LINK_METADATA_TYPE_NONE:
+		if o.Data != nil {
+			return NewAppError("LinkMetadata.IsValid", "model.link_metadata.is_valid.data_type.app_error", nil, "", http.StatusBadRequest)
+		}
+	case LINK_METADATA_TYPE_OPENGRAPH:
+		if o.Data == nil {
+			return NewAppError("LinkMetadata.IsValid", "model.link_metadata.is_valid.data.app_error", nil, "", http.StatusBadRequest)
+		}
+
+		if _, ok := o.Data.(*opengraph.OpenGraph); !ok {
+			return NewAppError("LinkMetadata.IsValid", "model.link_metadata.is_valid.data_type.app_error", nil, "", http.StatusBadRequest)
+		}
+	default:
+		return NewAppError("LinkMetadata.IsValid", "model.link_metadata.is_valid.type.app_error", nil, "", http.StatusBadRequest)
+	}
+
+	return nil
+}
+
+// FixData converts o.Data from JSON into properly structured data. This is intended to be used after deserializing
+// a LinkMetadata object that has been stored in the database.
+func (o *LinkMetadata) FixData() error {
```

Comment:
`FixData` is a generic name that can means anything, actually looks more like a some kind of correction in the stored data. Myabe the name could be something like `DeserializeDataToConcreteType` or something like that.

Reply:
Yeah, that's a good idea. Since I wasn't able to make it automatic, having a clearer method name is much better